modules/core/src/main/scala/sangria/execution/ValueCoercionHelper.scala

```diff
@@ -580,107 +580,126 @@ class ValueCoercionHelper[Ctx](
     nodeLocation.toList ++ firstValue.toList
   }
 
-  def isValidValue[In](tpe: InputType[_], input: Option[In])(implicit
-      um: InputUnmarshaller[In]): Vector[Violation] = (tpe, input) match {
-    case (OptionInputType(ofType), Some(value)) if um.isDefined(value) =>
-      isValidValue(ofType, Some(value))
-    case (OptionInputType(_), _) => Vector.empty
-    case (_, None) => Vector(NotNullValueIsNullViolation(sourceMapper, Nil))
-
-    case (ListInputType(ofType), Some(values)) if um.isListNode(values) =>
-      um.getListValue(values)
-        .toVector
-        .flatMap(v =>
-          isValidValue(
-            ofType,
-            v match {
-              case opt: Option[In @unchecked] => opt
-              case other => Option(other)
-            }).map(ListValueViolation(0, _, sourceMapper, Nil)))
-
-    case (ListInputType(ofType), Some(value)) if um.isDefined(value) =>
-      isValidValue(
-        ofType,
-        value match {
-          case opt: Option[In @unchecked] => opt
-          case other => Option(other)
-        }).map(ListValueViolation(0, _, sourceMapper, Nil))
-
-    case (objTpe: InputObjectType[_], Some(valueMap)) if um.isMapNode(valueMap) =>
-      val unknownFields = um.getMapKeys(valueMap).toVector.collect {
-        case f if !objTpe.fieldsByName.contains(f) =>
-          UnknownInputObjectFieldViolation(
-            SchemaRenderer.renderTypeName(objTpe, true),
-            f,
-            sourceMapper,
-            Nil)
-      }
+  def isValidValue[In](inputType: InputType[_], input: Option[In], errorsLimit: Option[Int])(
+      implicit um: InputUnmarshaller[In]): Vector[Violation] = {
+
+    // keeping track of the number of errors
+    var errors = 0
```

Comment:
Using mutable variable was the only viable option here to avoid making huge refactoring and keeping the same signature of the original method (by returning Vector[Violation]).

The alternative 1 was to use ListBuffer[Violation] but it would require making signature changes (probably just to return Unit) + it didn't work well with `.map(ListValueViolation(0, _, sourceMapper, Nil))` if we I did it.

The alternative 2 was to propagate the number of errors back to the recursion function but I didn't make it work properly 🙂